From d8d604c6e7dcfae8aa51d16d1d0033b2d9e132e9 Mon Sep 17 00:00:00 2001
From: Aloka Dixit <quic_alokad@quicinc.com>
Date: Fri, 10 Sep 2021 14:59:32 -0700
Subject: [PATCH 10/15] mbssid: DTIM period configuration for EMA AP

Set the DTIM periods of non-transmitted profiles equal to the EMA
profile periodicity if those are not a multiple of the latter
already as recommended in IEEE P802.11ax/D8.0, October 2020,
Multiple BSSID configuration examples, AA.1 Introduction.

Signed-off-by: Aloka Dixit <quic_alokad@quicinc.com>
---
 src/ap/ieee802_11.c | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7777,7 +7777,7 @@ size_t hostapd_eid_mbssid_len(struct hos
 
 static u8 * hostapd_eid_mbssid_elem(struct hostapd_data *hapd, u8 *eid, u8 *end,
 				    u32 frame_type, u8 max_bssid_indicator,
-				    size_t *bss_index)
+				    size_t *bss_index, u8 elem_count)
 {
 	struct hostapd_data *tx_bss = hostapd_mbssid_get_tx_bss(hapd);
 	size_t i;
@@ -7818,6 +7818,9 @@ static u8 * hostapd_eid_mbssid_elem(stru
 		if (frame_type == WLAN_FC_STYPE_BEACON) {
 			*eid++ = 3;
 			*eid++ = i;
+			if (hapd->iconf->ema &&
+			    (conf->dtim_period % elem_count))
+				conf->dtim_period = elem_count;
 			*eid++ = conf->dtim_period;
 			*eid++ = 0xFF;
 		} else {
@@ -7903,7 +7906,7 @@ u8 * hostapd_eid_mbssid(struct hostapd_d
 		}
 		eid = hostapd_eid_mbssid_elem(hapd, eid, end, frame_type,
 					      hostapd_max_bssid_indicator(hapd),
-					      &bss_index);
+					      &bss_index, elem_count);
 	}
 	return eid;
 }
